tests/unit/test_commands.py

```diff
@@ -822,56 +829,72 @@ def test_parse_args_upgrade(args, expected_CLIargs):
     assert parsed_args == expected_CLIargs
 
 
-@pytest.mark.parametrize(
-    "args",
-    [
-        ["upgrade", "hypervisors", "--machine 1", "--az 2"],
-        ["upgrade", "hypervisors", "--az 1", "-m 2"],
-    ],
-)
-def test_parse_args_hypervisors_exclusive_options(args):
+@patch("cou.commands.argparse.ArgumentParser.error")
+def test_parse_args_hypervisors_exclusive_options(mock_error):
     """Test parsing mutually exclusive hypervisors specific options."""
-    with pytest.raises(SystemExit, match="2"):
-        commands.parse_args(args)
+    mock_error.side_effect = SystemExit
+    with pytest.raises(SystemExit):
+        commands.parse_args(["upgrade", "hypervisors", "--machine", "1", "--az", "2"])
+    mock_error.assert_called_once_with(
+        "argument --availability-zone/--az: not allowed with argument --machine/-m"
+    )
+
+
+@patch("cou.commands.argparse.ArgumentParser.error")
+def test_parse_args_hypervisors_exclusive_options_reverse_order(mock_error):
+    """Test parsing mutually exclusive hypervisors specific options (reverse order)."""
+    mock_error.side_effect = SystemExit
+    with pytest.raises(SystemExit):
+        commands.parse_args(["upgrade", "hypervisors", "--az", "1", "-m", "2"])
+    mock_error.assert_called_once_with(
+        "argument --machine/-m: not allowed with argument --availability-zone/--az"
+    )
 
 
 @pytest.mark.parametrize(
     "args",
     [
+        ["foo"],
         ["upgrade", "--archive-batch-size", "asdf"],
         ["plan", "--archive-batch-size", "asdf"],
         ["upgrade", "--archive-batch-size", "-4"],
         ["plan", "--archive-batch-size", "-4"],
         ["upgrade", "--archive-batch-size", "0"],
         ["plan", "--archive-batch-size", "0"],
         ["plan", "--purge_before", "2000-01-02"],
-        ["plan", "--purge_before", "2000-01-02 03:04"],
-        ["plan", "--purge_before", "2000-01-02 03:04:05"],
+        ["plan", "--purge_before", "2000-01-0203:04"],
+        ["plan", "--purge_before", "2000-01-0203:04:05"],
         ["upgrade", "--skip-apps", "vault keystone"],
         ["plan", "--skip_apps", "vault keystone"],
     ],
 )
-def test_parse_invalid_args(args):
+@patch("cou.commands.argparse.ArgumentParser.error")
+def test_parse_invalid_args(mock_error, args):
     """Generic test for various invalid sets of args."""
-    with pytest.raises(SystemExit, match="2"):
+    mock_error.side_effect = SystemExit
+    with pytest.raises(SystemExit):
         commands.parse_args(args)
+    mock_error.assert_called()
```

Comment:
Why do you need to mock error also here? What value we will get from it like this?

Reply:
Here, I guess no real value. But at least it gives us opportunity to add value.

Reply:
For me we are simply losing value here, since we expect to raise something and just mocked it. If you want to check that it was called, I will rather use `wraps` so we can spy on object.

Reply:
> I will rather use wraps so we can spy on object.

This sounds interesting, could you explain more about what you mean or show links? I'm not aware of that.

Reply:
You can check it [here](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock). The wraps can be used also with patch, so you will be able to check that it was called without mocking the output.

Reply:
Nice, thanks! I'll update the patching to use that. :+1: :+1:

Reply:
Hmm it doesn't look straight forward for using this for a class method, but I'll give it a shot.

Reply:
@rgildein have you used this kind of mock wrapping before? I can't get it to work.

Reply:
```python
@patch("cou.commands.argparse.ArgumentParser.error", wraps=ArgumentParser.error, autospec=True)
def test_parse_args_quiet_verbose_exclusive(mock_error, args):
    """Test that quiet and verbose options are mutually exclusive."""
    with pytest.raises(SystemExit, match="2"):
        commands.parse_args(args)

    mock_error.assert_called_once_with(
        ANY, "argument --verbose/-v: not allowed with argument --quiet/-q"
    )
```

Reply:
ooh nice, `autospec` was what I was missing. That's kinda magic. Thanks for that :)

Reply:
Updated, ready for review again. :)

Reply:
@chanchiwai-ray please read this thread: https://github.com/canonical/charmed-openstack-upgrader/pull/580#discussion_r1794772211

Reply:
@rgildein are you aware of any differences in the mocking lib between python 3.10 and 3.12 that could cause this to work on 3.12 but not 3.10?

Reply:
Ah, it's a bug in cpython: https://github.com/python/cpython/issues/75988

Reply:
Looks like it was backported to 3.11 and 3.12, but not 3.10. So we can't use this method here unfortunately unless we can find a workaround.

I propose dropping back to the earlier method: no `wraps`, and manually adding the SystemExit sideeffect.

Reply:
cc @rgildein @chanchiwai-ray ^

Reply:
Resolving as I think the latest solution is the most reasonable option for now, given the comments and constraints we have in python 3.10. We can refactor later. :)

Reply:
@rgildein please let me know if you know of any better method? Otherwise we can wait until we switch to python >= 3.11 to use `wraps` with patch here.